server/service/src/programs/program_indicator/query.rs

```diff
@@ -0,0 +1,231 @@
+use repository::{
+    IndicatorColumnRow, IndicatorColumnRowRepository, IndicatorLineRow, IndicatorLineRowRepository,
+    IndicatorValueType, Pagination, ProgramIndicatorFilter, ProgramIndicatorRepository,
+    ProgramIndicatorRow, ProgramIndicatorSort, RepositoryError, StorageConnection,
+};
+
+#[derive(Clone, serde::Serialize)]
+pub enum ColumnValue {
+    Text(String),
+    Number(f64),
+}
+
+#[derive(Clone, serde::Serialize)]
+pub enum ValueType {
+    String,
+    Number,
+}
+
+#[derive(Clone, serde::Serialize)]
+pub struct IndicatorColumn {
+    pub header: String,
+    pub r#type: ValueType,
+    pub value: ColumnValue,
+}
+
+#[derive(Clone, serde::Serialize)]
+pub struct IndicatorLine {
+    pub name: String,
+    pub code: String,
+    pub value: Vec<IndicatorColumn>,
+}
+
+#[derive(serde::Serialize, Clone)]
+pub struct ProgramIndicator {
+    pub id: String,
+    pub program_id: String,
+    pub code: String,
+    pub lines: Vec<IndicatorLine>,
+}
+
+pub fn program_indicator(
+    connection: &StorageConnection,
+    filter: ProgramIndicatorFilter,
+) -> Result<Option<ProgramIndicator>, RepositoryError> {
+    let indicator = ProgramIndicatorRepository::new(&connection)
+        .query_by_filter(filter)?
+        .pop();
+
+    if let Some(indicator) = indicator {
+        // grafind all relevant lines
+        let all_indicator_line_rows = IndicatorLineRowRepository::new(&connection)
+            .find_many_by_indicator_id(indicator.id.clone())?;
+
+        // find all relevant columns
+        let all_indicator_column_rows = IndicatorColumnRowRepository::new(&connection)
+            .find_many_by_indicator_id(indicator.id.clone())?;
+
+        let program_indicator = ProgramIndicator::from_domain(
+            indicator,
+            all_indicator_line_rows,
+            all_indicator_column_rows,
+        );
+
+        Ok(Some(program_indicator))
+    } else {
+        Ok(None)
+    }
+}
+
+pub fn program_indicators(
+    connection: &StorageConnection,
+    pagination: Pagination,
+    sort: Option<ProgramIndicatorSort>,
+    filter: Option<ProgramIndicatorFilter>,
+) -> Result<Vec<ProgramIndicator>, RepositoryError> {
+    let indicators =
+        ProgramIndicatorRepository::new(&connection).query(pagination, filter, sort)?;
+
+    let indicator_ids: Vec<String> = indicators
```

Comment:
HashMaps would be a great way to organise the indicators with their rows and columns (;  
Would save you time remapping too!